### PR TITLE
Add chapter: Introduction to Node.js

### PR DESCRIPTION
**Pull request by rachx**

### CODE REVIEW

File: contents/node/nodejs.md
Comment (tshradheya) [inline contents/node/nodejs.md]: Shouldn't this be `node server.js` ?

File: contents/node/nodejs.md
Comment (nicholaschuayunzhi) [inline contents/node/nodejs.md]: *an* event *happens*

File: contents/node/nodejs.md
Comment (nicholaschuayunzhi) [inline contents/node/nodejs.md]: I believe NPM is stylized as npm

File: contents/node/nodejs.md
Comment (nicholaschuayunzhi) [inline contents/node/nodejs.md:152]: perhaps you could illustrate how to install these modules and also add them as dependencies. ie `$ npm install bluebird --save`

File: contents/node/nodejs.md
Comment (nicholaschuayunzhi) [reply contents/node/nodejs.md]: Should we standardize placing a `$` for all terminal commands?

**Review by nicholaschuayunzhi — COMMENTED** [review]
Hi @rachx, have left some comments about your chapter. Overall I feel the chapter is rather comprehensive.

File: contents/node/nodejs.md
Comment (mauris) [inline contents/node/nodejs.md]: seems like there's no JavaScript intro. would be good if someone can look into it too :D

File: contents/node/nodejs.md
Comment (rachx) [reply contents/node/nodejs.md:152]: Amended the last commit and added this in the dependencies section and a few lines up

File: contents/node/nodejs.md
Comment (nicholaschuayunzhi) [inline contents/node/nodejs.md]: The quotation seems a bit off. Maybe you could do without it?

*Node.js* is a JavaScript runtime built on Chrome’s V8 JavaScript engine. It uses an event-driven, non-blocking I/O model that makes it lightweight and efficient.

File: contents/node/nodejs.md
Comment (nicholaschuayunzhi) [inline contents/node/nodejs.md]: Should add `$` to all the terminal commands in this line.

**Review by nicholaschuayunzhi — CHANGES_REQUESTED** [review]
Hi @rachx, had a second look and i felt this intro was very comprehensive. Just requesting some minor changes.

File: contents/node/nodejs.md
Comment (rachx) [reply contents/node/nodejs.md]: I thought it was weird to add it inline since it says runs x in the command shell

**Comment by rachx** [comment]
@tshradheya @nicholaschuayunzhi Thanks for the review! Can you review it again and approve if it is okay?

**Review by nicholaschuayunzhi — APPROVED** [review]

**Comment by tshradheya** [comment]
I will probably not approve this as I haven't done a proper review. And also I might require more time to actually do the complete review, so feel free to ignore my review and move on to asking others for further reviews :P

**Comment by rachx** [comment]
Prof @damithc, may I have your review?

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Enough to mention that it is simple download and launch.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Contents in this section up to this point are fine. Although you are giving instructions, they can serve as a general description of how easy it is to load dependencies.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Replace this part by a description of how useful the `package.json` mechanism is.

**Review by damithc — CHANGES_REQUESTED** [review]
Good start. Can we reframe some of the sections to remove the appearance of a step-by-step tutorial? There are plenty of such tutorials out there. Instead, describe virtues of node to motivate the reader. e.g.,  Rather than how to get started on node.js, frame it as 'see how easy to get started'.  It is fine to give concrete examples of commands/code etc, but they should be to prove your point rather than to teach how to use. I have added a few comments to this effect. You can follow a similar approach for the rest of the chapter. If not clear, ask me during tomorrow session.

**Comment by rachx** [comment]
I removed some tutorial steps/explanation and re-frame some of the content to how is it easy to use/useful.

@nicholaschuayunzhi, Prof @damithc , can I have more suggestions?

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: >...blocking - subsequent...
Should this be double dash? Not sure.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Best to mention the name `npm` here.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: unexplained term `npm`

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: The section on NPM still feels like you are trying to teach how to use it instead of explaining how useful it is.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: This example serves to motivate the reader to adopt node, by showing how easy it is to use. If possible, move this section higher up (under the `Getting started` section?). It will motivate the reader to keep reading the remainder of the chapter.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Does this chapter assumes the reader is familiar with JS? if so, mention that. If not, the chapter can be read by a wider audience (e.g., to motivate people to learn JS in order to use a cool tool like node.js) but then you need to tweak the text accordingly to avoid alienating reading who are not familiar with JS.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Can you give at least a few examples and point the reader to other articles to find more. Simply pointing to other articles feels like cheating :-p

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Event driven and non-blocking can be bullet points rather than headings.
>* **Event-driven** : When an event happens e.g. an I/O event is complete, the event handler/callback function will be enqueued in a queue where it will be subsequently scheduled to run by the [event loop](https://medium.com/the-node-js-collection/what-you-should-know-to-really-understand-the-node-js-event-loop-and-its-metrics-c4907b19da4c).

**Review by damithc — CHANGES_REQUESTED** [review]
Moving in the right direction. Added a few comments. Tweak the content further to make it feel less like teaching/tutorial.

**Comment by rachx** [comment]
Prof @damithc, I have made some changes

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: May br remove this heading? The content can fit inside the `Getting Started` heading.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: rode -> role?

**Review by damithc — CHANGES_REQUESTED** [review]
Minor nits. We can merge after that.

**Comment by damithc** [comment]
For your reference, another way to structure this chapter is (for better high-level structure):
# Overview
 What it is ... 

# Benefits
Why use it ...
## Benefit 1: Easy to get started
## Benefit 2: 
## Benefit 3: Easy dependency management
NPM is described here
## Benefit 4: Support for better code organization
Module system is described here

# Use cases
When to use and when not to use ...

**Comment by rachx** [comment]
Prof @damithc , I restructured the chapter as well

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Nice work. Any chance you can reconcile the list above with the sections below. i.e., to match the bullet list with the sub-sections, as both are benefits? One option is to convert the bullet list to sub sections too. i.e., have 9 benefits.

**Review by damithc — COMMENTED** [review]
Nice! just one more comment.

**Comment by rachx** [comment]
Prof @damithc, updated!

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: This reads better?
> [_About Asynchronous Programming_ section of the _You Don't Know JS Guide_]()

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Why not move the _non-blocking_ explanation here, so that this section is not so sparse?

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Similarly, you can move the _event-driven_ explanation here (if it fits)?

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: You may want to place this after benefit 6, by which time the reader has a better appreciation of npm. In fact you can put the npm-related benefits close to each other and sequence them in a way that later ones build on earlier ones.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md:201]: Is there a way a reader can tryout node without installing it? like an online platform? Just curious.

**Review by damithc — COMMENTED** [review]
LGTM. Added few comments. I'll leave you to decide if you want to follow them, as you are the content expert. Let me know when you are done so that I can merge.

File: contents/node/nodejs.md
Comment (rachx) [reply contents/node/nodejs.md:201]: I just went to check and there is [RunKit]. It should work in the browser and it has all the modules/versions pre-installed.

Prof @damithc, the rest of the content has been updated.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: May be rephrase to sound like a benefit right away? e.g., `No threading overhead`?
Alternatively, combine 1 and 2 `Better performance without the costs of multi-threading`.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: I would put this as the first benefit. Although it is not a big benefit, it gives the reader a concrete sense of what Node.js is before talking about more abstract benefits.

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Mention where it fits into the Web technology landscape? If not, the reader might assume it is a front-end technology. Giving a few example scenarios of how node is used (for an architecture point of view) can be useful too. e.g., a web app written using node.js, a non-web app using node etc. Can be real or hypothetical

Comment (damithc) [inline contents/node/nodejs.md]: This sentence seems to be from https://nodejs.org. cite? e.g.,
>*Node.js* is a JavaScript runtime built on Chrome’s V8 JavaScript engine. Node.js uses an event-driven, non-blocking I/O model that makes it lightweight and efficient. <sub>--https://nodejs.org</sub>

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: we don't use Node.js in the heading of other benefits. Remove it here as well, for consistency?

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: Remove `Node` from heading here too?

**Review by damithc — CHANGES_REQUESTED** [review]
A few changes suggested, mainly to help ease the reader into the chapter.

**Comment by rachx** [comment]
Updated

File: contents/node/nodejs.md
Comment (damithc) [inline contents/node/nodejs.md]: `=` → `-` ?

**Review by damithc — APPROVED** [review]
Possibly a typo to fix. You can go ahead and merge it yourself after that. Do a 'squash and merge'. Good job doing all the improvements along the way 💯

**Comment by damithc** [comment]
@rachx looks like this is not linked from the main Table of Contents. Can you do a quick PR to add it there?

**Pull request merged**
